include/kernels/dslash_ndeg_twisted_clover.cuh

```diff
@@ -0,0 +1,112 @@
+      }
+  };
+  
+  template <int nParity, bool dagger, bool xpay, KernelType kernel_type, typename Arg>
+    struct nDegTwistedClover : dslash_default {
+    
+    Arg &arg;
+    constexpr nDegTwistedClover(Arg &arg) : arg(arg) {}
+    static constexpr const char *filename() { return KERNEL_FILE; } // this file name - used for run-time compilation
+    
+    /**
+       @brief Apply the non-degenerate twisted-clover dslash
+       out(x) = M*in = A(x)*in(x) + a * D * in + (1 + i*b*gamma_5*tau_3 + c*tau_1)*in
+       Note this routine only exists in xpay form.
+    */
+    __device__ __host__ inline void operator()(int idx, int flavor, int parity)
+    {
+      typedef typename mapper<typename Arg::Float>::type real;
+      typedef ColorSpinor<real, Arg::nColor, 4> Vector;
+      typedef ColorSpinor<real, Arg::nColor, 2> HalfVector;
+      
+      bool active
+        = kernel_type == EXTERIOR_KERNEL_ALL ? false : true; // is thread active (non-trival for fused kernel only)
+      int thread_dim;                                        // which dimension is thread working on (fused kernel only)
+
+      // FIXME: do we actually need two of these?
+      // checking getCoords, the 's' parameter (where we pass flavor) simply sets coord.s
+      // such that we likely don't need to do this, as the 4D coords are unchanged
+      auto clover_coord = getCoords<QUDA_4D_PC, kernel_type>(arg, idx, 0, parity, thread_dim);
+      auto coord = getCoords<QUDA_4D_PC, kernel_type>(arg, idx, flavor, parity, thread_dim);
+      
+      const int my_spinor_parity = nParity == 2 ? parity : 0;
+      Vector out;
+      
+      // defined in dslash_wilson.cuh
+      applyWilson<nParity, dagger, kernel_type>(out, arg, coord, parity, idx, thread_dim, active);
+
+      int my_flavor_idx = coord.x_cb + flavor * arg.dc.volume_4d_cb;
+
+      if (kernel_type == INTERIOR_KERNEL) {
+        // apply the chiral and flavor twists
+        // use consistent load order across s to ensure better cache locality
+        Vector x0 = arg.x(coord.x_cb + 0 * arg.dc.volume_4d_cb, my_spinor_parity);
+        Vector x1 = arg.x(coord.x_cb + 1 * arg.dc.volume_4d_cb, my_spinor_parity);
+
+        Vector x = flavor == 0 ? x0 : x1;
+        x.toRel(); // switch to chiral basis
+        
+        Vector tmp;
+#pragma unroll
+        for (int chirality = 0; chirality < 2; chirality++) {
+          constexpr int n = Arg::nColor * Arg::nSpin / 2;
+          HMatrix<real, n> A = arg.A(clover_coord.x_cb, parity, chirality);
+          HalfVector x_chi = x.chiral_project(chirality);
+          HalfVector Ax_chi = A * x_chi;
+          // i * mu * gamma_5 * tau_3
+          const complex<real> b(0.0, (chirality^flavor) == 0 ? static_cast<real>(arg.b) : -static_cast<real>(arg.b));
+          Ax_chi += b * x_chi;
+          tmp += Ax_chi.chiral_reconstruct(chirality);
+        }
+
+        tmp.toNonRel();
+        // c * tau_1
+        tmp += (flavor == 0 ? arg.c * x1 : arg.c * x0);
```

Comment:
@urbach Unlike `b`, The normalisation of `c` should be 1.0, right? It is added to `tmp` as a full `Vector` rather than separately to each chirality.

Wanted to make sure that we don't have a misunderstanding as your write in the Wiki that also `c` shoud have this factor:

https://github.com/qcdcode/quda/wiki/ndeg-twisted-clover-kernel-and-testing/_compare/c8cb509073ebe5aad563ff8768bd851f996fb884...33d7f42d8bd067a0b805f7a360c8df5de4b75044#diff-b014db1857a953f6c396fc04597381a347d941d3463fe4e3d318d840eb0ee482R9

Reply:
Similarly, `a` should also not be scaled.

Reply:
I agree with you. Didn't I say it is only a facto 1/2 in the chiral basis?

Reply:
I have clarified this in the docu.